File: src/Exporter.NX.Nx/ComponentExtractor.cs
// SPDX-License-Identifier: GPL-2.0-only
using System;
using System.Collections.Generic;
using System.Runtime.CompilerServices;
using NXOpen;
using NXOpen.Assemblies;
using Oblikovati.Exporter.NX.Model;

namespace Oblikovati.Exporter.NX.Nx
{
    /// <summary>
    /// Walks an NX component tree into IR occurrences. Each component becomes an
    /// <see cref="NxOccurrence"/> referencing the NX-neutral document of its prototype part
    /// (extracted via the supplied part extractor, deduped by prototype so a part placed N
    /// times exports once). A sub-assembly component recurses; a leaf component extracts its
    /// part. The orientation's columns are the placed X/Y/Z axes (NX convention), so the
    /// row-major 3x3 the IR carries is filled column-wise.
    /// </summary>
    public sealed class ComponentExtractor
    {
        private readonly Func<Part, NxDocument> _extractPart;
        private readonly Dictionary<Part, NxDocument> _docs = new Dictionary<Part, NxDocument>(ReferenceComparer.Instance);

        public ComponentExtractor(Func<Part, NxDocument> extractPart)
        {
            _extractPart = extractPart ?? throw new ArgumentNullException(nameof(extractPart));
        }

        public NxOccurrence Occurrence(Component component)
        {
            component.GetPosition(out Point3d origin, out Matrix3x3 orientation);
            return new NxOccurrence
            {
                Name = component.DisplayName,
                Component = DocumentFor(component.Prototype),
                Position = new[] { origin.X, origin.Y, origin.Z },
                Rotation = RowMajor(orientation),
            };
        }

        private NxDocument DocumentFor(Part prototype)
        {
            if (_docs.TryGetValue(prototype, out NxDocument? existing))
            {
                return existing;
            }

            Component[] children = prototype.ComponentAssembly.RootComponent?.GetChildren() ?? Array.Empty<Component>();
            NxDocument doc = children.Length > 0 ? Subassembly(prototype, children) : _extractPart(prototype);
            _docs[prototype] = doc;
            return doc;
        }

        private NxDocument Subassembly(Part prototype, Component[] children)
        {
            var doc = new NxDocument
            {
                DisplayName = prototype.Leaf,
                Kind = NxDocumentKind.Assembly,
                LengthUnit = prototype.PartUnits == PartUnits.Inches ? "in" : "mm",
            };
            foreach (Component child in children)
            {
                doc.Occurrences.Add(Occurrence(child));
            }

            return doc;
        }

        // Columns of the NX orientation are the placed axes; lay them out row-major.
        private static double[] RowMajor(Matrix3x3 m) => new[]
        {
            m.Xx, m.Yx, m.Zx,
            m.Xy, m.Yy, m.Zy,
            m.Xz, m.Yz, m.Zz,
        };

        private sealed class ReferenceComparer : IEqualityComparer<Part>
        {
            public static readonly ReferenceComparer Instance = new ReferenceComparer();

            public bool Equals(Part? x, Part? y) => ReferenceEquals(x, y);

            public int GetHashCode(Part obj) => RuntimeHelpers.GetHashCode(obj);
        }
    }
}

File: python/oblikovati_exporter_nx/nx/component_extractor.py
# SPDX-License-Identifier: GPL-2.0-only
"""Walks an NX component tree into IR occurrences.

Each component becomes an NxOccurrence referencing the NX-neutral document of its
prototype part (extracted via the supplied part extractor, deduped by prototype so a part
placed N times exports once). A sub-assembly component recurses; a leaf component extracts
its part. The orientation's columns are the placed X/Y/Z axes (NX convention), so the
row-major 3x3 the IR carries is filled column-wise. ``GetPosition()`` returns
(origin, orientation) as a tuple in NXOpen Python.

UNVERIFIED: needs a real NX session.
"""
from __future__ import annotations

from typing import Callable, Dict

from ..model.document import NxDocument, NxDocumentKind, NxOccurrence


class ComponentExtractor:
    def __init__(self, extract_part: Callable[[object], NxDocument]) -> None:
        self._extract_part = extract_part
        self._docs: Dict[int, NxDocument] = {}  # keyed by prototype .Tag

    def occurrence(self, component) -> NxOccurrence:
        origin, orientation = component.GetPosition()
        return NxOccurrence(
            name=component.DisplayName,
            component=self._document_for(component.Prototype),
            position=[origin.X, origin.Y, origin.Z],
            rotation=_row_major(orientation),
        )

    def _document_for(self, prototype) -> NxDocument:
        existing = self._docs.get(prototype.Tag)
        if existing is not None:
            return existing

        root = prototype.ComponentAssembly.RootComponent
        children = list(root.GetChildren()) if root is not None else []
        doc = self._subassembly(prototype, children) if children else self._extract_part(prototype)
        self._docs[prototype.Tag] = doc
        return doc

    def _subassembly(self, prototype, children) -> NxDocument:
        doc = NxDocument(
            display_name=prototype.Leaf,
            kind=NxDocumentKind.ASSEMBLY,
            length_unit=_length_unit_of(prototype),
        )
        for child in children:
            doc.occurrences.append(self.occurrence(child))
        return doc


# Columns of the NX orientation are the placed axes; lay them out row-major.
def _row_major(m) -> list:
    return [
        m.Xx, m.Yx, m.Zx,
        m.Xy, m.Yy, m.Zy,
        m.Xz, m.Yz, m.Zz,
    ]


def _length_unit_of(part) -> str:
    import NXOpen

    return "in" if part.PartUnits == NXOpen.Part.Units.Inches else "mm"
